fix(core): memory leak in event listeners inside embedded views

crisbeto opened the pull request:
When we have an event listener inside an embedded view, we generate a `restoreView` call which saves the view inside of the LFrame. The problem is that we don't clear it until it gets overwritten which can lead to memory leaks.

These changes rework the generated code in order to generate a `resetView` call which will clear the view from the LFrame.

Fixes #42848.

JoostK commented on packages/compiler/src/compiler_util/expression_converter.ts:
```diff
@@ -54,13 +55,16 @@ export class ConvertActionBindingResult {
      */
     // TODO(misko): remove this hack once we no longer support ViewEngine.
-    this.render3Stmts = stmts.map((statement: o.Statement) => {
+    stmts.forEach((statement: o.Statement) => {
```
nit: a `for..of` loop would be preferred here

JoostK commented on packages/compiler-cli/test/compliance/test_cases/r3_view_compiler_template/nested_template_context_template.js:
```diff
@@ -8,7 +8,9 @@ function MyComponent_ul_0_li_1_div_1_Template(rf, ctx) {
       const $outer$ = $i0$.ɵɵnextContext().$implicit;
       const $myComp$ = $i0$.ɵɵnextContext();
-      return $myComp$.onClick($outer$, $middle$, $inner$);
+      const $returnValue$ = $myComp$.onClick($outer$, $middle$, $inner$);
+      $i0$.ɵɵresetView();
+      return $returnValue$;
```
Did you consider something like:

```ts
       return $i0$.ɵɵleaveListener($myComp$.onClick($outer$, $middle$, $inner$));
```

<sup>(open to bike-shedding the instruction name)</sup>

It would save some bytes.

crisbeto replied:
Interesting idea, I hadn't considered it. My only concern is the naming of the instruction.

AndrewKushnir replied:
I was also thinking if it might make sense to introduce a new instruction altogether (called `ɵɵembeddedViewListener` for example) that would take care of setting up global state as needed (call `restoreView` at the beginning and `resetView` at the end) and invoke a listener?

crisbeto replied:
We could go down the route of `ɵɵembeddedViewListener`, but there are some cases where we use the return value of `restoreView` within the listener https://github.com/angular/angular/blob/47f87584b58efcf34005e02aa52c81d1655ed4f2/packages/compiler-cli/test/compliance/test_cases/r3_view_compiler_template/nested_template_context_many_bindings_template.js#L6.

AndrewKushnir replied:
> We could go down the route of ɵɵembeddedViewListener, but there are some cases where we use the return value of restoreView within the listener

We can probably add a new instruction here to return the context of the current `contextLView` from the global state.

AndrewKushnir replied:
@crisbeto sounds good 👍
Just wanted to mention that I brought it up as an idea and I'm ok if we implement it differently :)

petebacondarwin replied:
I do like the idea of moving the `restoreView()` and `releaseView()` calls into a new instruction:

```ts
export function ɵɵembeddedListener(
    eventName: string, viewToRestore: OpaqueViewState, listenerFn: (e?: any) => any,
    useCapture?: boolean, eventTargetResolver?: GlobalTargetResolver): typeof ɵɵlistener {
  const lView = getLView();
  const tView = getTView();
  const tNode = getCurrentTNode()!;
  instructionState.lFrame.contextLView = viewToRestore;
  try {
    listenerInternal(
        tView, lView, lView[RENDERER], tNode, eventName, listenerFn, !!useCapture,
        eventTargetResolver);
  } finally {
    instructionState.lFrame.contextLView = null;
  }
  return ɵɵembeddedListener;
}
```

and then adding a new instruction to access the current view:

```ts
export function ɵɵgetLView() {
  return getContextLView();
}
```
```

JoostK replied:
The thing is that the context needs to be restored when the listener is invoked, not when it's registered.

petebacondarwin replied:
Oops! Yes - clearly it is too late for me. I need to get to sleep.

jessicajaniuk commented:
@crisbeto Did you still want to land this PR?

crisbeto commented:
Yes, I think that it's still valid.

crisbeto committed:
fix(core): memory leak in event listeners inside embedded views

When we have an event listener inside an embedded view, we generate a `restoreView` call which saves the view inside of the LFrame. The problem is that we don't clear it until it gets overwritten which can lead to memory leaks.

These changes rework the generated code in order to generate a `resetView` call which will clear the view from the LFrame.

Fixes #42848.

crisbeto commented:
I've reworked the PR based on the discussion above.

JoostK commented on packages/compiler/src/render3/view/template.ts:
```diff
@@ -81,14 +81,32 @@ export function prepareEventListenerParameters(
+    const lastStatement = statements[statements.length - 1];
+    if (lastStatement instanceof o.ReturnStatement) {
+      statements[statements.length - 1] =
+          new o.ReturnStatement(invokeInstruction(null, R3.resetView, [lastStatement.value]));
```
Should this inherit the source span of the original statement?

```suggestion
          new o.ReturnStatement(invokeInstruction(lastStatement.sourceSpan, R3.resetView, [lastStatement.value]));
```

crisbeto replied:
Changed, but I think that we actually want `lastStatement.value.sourceSpan`.

JoostK reviewed (approved):
LGTM with one comment

AndrewKushnir reviewed (approved):
LGTM 👍 

As a followup cleanup, I'd propose to revisit function names that set and unset the context LView. Current names `ɵɵrestoreView` and `ɵɵresetView` are quite confusing. It'd be great to have "cooperating" names, like "enterView"+"leaveView" or "setContextView"+"unsetContextView".

AndrewKushnir reviewed (approved):
Reviewed-for: size-tracking

jessicajaniuk reviewed (approved):
reviewed-for: size-tracking

crisbeto committed:
fixup! fix(core): memory leak in event listeners inside embedded views

jelbourn reviewed (approved):
LGTM

Reviewed-for: size-tracking

atscott commented:
merge assistance: CI looks stuck. This has all necessary approvals.

dylhunn commented:
This PR has merge conflicts with patch. I will merge into `master` for now; let me know if you want to rebase and patch port!

dylhunn commented:
This PR was merged into the repository by commit be161bef798ac5309df8390e3c2bba2cf84cb530.
